File: Leetcode/279.完全平方数.cpp
/*
 * @lc app=leetcode.cn id=279 lang=cpp
 *
 * [279] 完全平方数
 */

#include<iostream>
using namespace std;

// @lc code=start
class Solution {
public:
    int numSquares(int n) {
        // int rt = sqrt(n); // 下取整
        // 1^2 = 1
        // n = \sum a_i^2
        // n - a_j^2 = \sum a
        // f(n) = f(n-a_j^2) + 1(a_j^2), a_j varies from sqrt(n) to 1
        
        vector<int> f(n+1);
        for (int i = 1; i <= n; ++i) {
            int minn = INT_MAX;
            int cur_top = sqrt(i);
            for (int j = 1; j <= cur_top; ++j) {
                minn = min(minn, f[i - j * j]);
            }
            f[i] = minn + 1;
        }
        return f[n];

    }
};
// @lc code=end

File: Leetcode/279.完全平方数.py
#
# @lc app=leetcode.cn id=279 lang=python3
#
# [279] 完全平方数
#

import math

# @lc code=start
class Solution:
    def numSquares(self, n: int) -> int:
        f = [0] * (n+1)
        for i in range(1, n+1):
            cur = math.inf
            for j in range(1, math.ceil(math.sqrt(i)) + 1):
                if j * j > i: continue
                cur = min(cur, f[i-j*j])
            f[i] = cur + 1
        return f[n]
# @lc code=end

